consensus/src/backup/loader.rs

```diff
@@ -63,14 +64,29 @@ impl From<CodecError> for LoaderError {
     }
 }
 
-pub struct BackupLoader<H: Hasher, D: Data, S: Signature, R: Read> {
+#[async_trait]
+pub trait BackupReader {
```

Comment:
If possible it would be nice to not create additional traits here at all. Could you just use `futures::io::AsyncRead`?

Reply:
https://docs.rs/futures/latest/futures/io/trait.AsyncRead.html is a very inconvenient interface to work with on both sides, because it is meant to be used as general purpose IO trait, not just a single use case like here. It will probably get changed in the future as well when (if?) Rust developers figure out how to square the circle of all the requirements put on it: https://www.ncameron.org/blog/async-read-and-write-traits/

I don't see any benefit of using it. In this case core alephbft logic wants the implementation to handle a single simple operation so expressing it as such makes things easier.

Reply:
This is kinda a good argument for leaving it as the sync version of `Read`, implementing it for a buffer and reading the data into a buffer separately, in whatever way you want. In fact this is what we do in `aleph-node`: https://github.com/Cardinal-Cryptography/aleph-node/blob/9ecbc0d50b52acc90e9c33acc0fed3d64436fcc7/finality-aleph/src/party/backup.rs#L64.

That being said, this only works for `Read`, do you have a proposal for the writing part?

Reply:
using converting from sync `Read` to async Read is trivial using `AllowStdIo` in futures-io.